### PR TITLE
Add FilterCommand

### PR DESCRIPTION
**Pull request by ethanwangkangen**

### CODE REVIEW

File: src/main/java/seedu/academyassist/logic/commands/FilterCommand.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Creates a FilterCommand
+     * @param filterParam FilterParam.YEARGROUP or FilterParam.CLASS
+     */
+    public FilterCommand(FilterParam filterParam, Object filterValue) {
+        // Ensure that the filterParam and filterValue are not null using assertions
+        assert filterParam != null : "FilterParam should not be null";
+        assert filterValue != null : "FilterValue should not be null";
+        assert filterValue != null : "FilterValue should not be null";
```
Comment (cweijin) [inline src/main/java/seedu/academyassist/logic/commands/FilterCommand.java:43]: might be some unintended duplication here

File: src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        // Ensure that either the year group or subject prefix is present, but not both
+        boolean hasYearGroup = argMultimap.getValue(PREFIX_YEARGROUP).isPresent();
+        boolean hasSubject = argMultimap.getValue(PREFIX_SUBJECT).isPresent();
+
+        if (hasYearGroup && hasSubject) {
+            throw new ParseException("Please specify either a year group or a subject, not both.");
+        } else if (!hasYearGroup && !hasSubject) {
+            throw new ParseException("You must specify either a year group or a subject.");
+        }
```
Comment (cweijin) [inline src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java:45]: great that you considered all possible cases!

File: src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        String filterString;
+        Object filterValue;
+
+        if (hasYearGroup) {
+            // Extract the value after yg/
+            filterString = argMultimap.getValue(PREFIX_YEARGROUP).get();
+            // Call ParserUtil to parse year group
+            filterValue = ParserUtil.parseYearGroup(filterString);
+            return new FilterCommand(filterParam, filterValue);
+        } else { // hasSubject must be true
```
Comment (cweijin) [inline src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java:60]: thanks for including the comments for clarity!

File: src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
```
Comment (cweijin) [inline src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java:75]: it seems that this method appears in multiple parsers. Perhaps we could extract it into ParserUtil for reuse?

**Review by cweijin — COMMENTED** [review]
LGTM!

File: docs/UserGuide.md [hunk 5]
```diff
@@ -259,7 +259,7 @@ If you ever have questions about your data, just ask - we're always happy to hel
 ## FAQ
 
 **Q**: How do I transfer my data to another Computer?<br>
-**A**: Install the app in the other computer and overwrite the empty data file it creates with the file that contains 
+**A**: Install the app in the other computer and overwrite the empty data file it creates with the file that contains
```
Comment (JoanneHing) [inline docs/UserGuide.md:262]: Thanks for fixing all the whitespaces!

File: src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            // Extract the value after yg/
+            filterString = argMultimap.getValue(PREFIX_YEARGROUP).get();
+            // Call ParserUtil to parse year group
+            filterValue = ParserUtil.parseYearGroup(filterString);
+            return new FilterCommand(filterParam, filterValue);
+        } else { // hasSubject must be true
+            // Extract the value after s/
+            filterString = argMultimap.getValue(PREFIX_SUBJECT).get();
+            // Call ParserUtil to parse subject
+            filterValue = ParserUtil.parseSubject(filterString);
```
Comment (JoanneHing) [inline src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java:64]: Same as comment that WeiJin gave! The comments helps a lot in understanding the code!

**Review by JoanneHing — COMMENTED** [review]
Overall LGTM! Some test cases could be added to test the code!

**Review by cweijin — APPROVED** [review]
LGTM!

File: docs/UserGuide.md [hunk 7]
```diff
@@ -28,15 +28,15 @@ title: AcademyAssist User Guide
...
+Welcome to AcademyAssist, your ultimate solution for efficient student contact management in tuition centers.
+This desktop application is designed to streamline your administrative tasks, allowing you to focus more
 on what truly matters - educating and nurturing young minds.
 
-AcademyAssist combines the power of a **Command Line Interface (CLI) with the user-friendly aspects of a Graphical 
-User Interface (GUI).** This hybrid approach ensures that you can manage your student database with 
-lightning-fast efficiency while still enjoying the visual benefits of modern software design. Whether you're a 
-seasoned administrator or new to digital management systems, AcademyAssist is intuitive enough for beginners yet 
-powerful enough for experts. With features like quick student addition, easy information retrieval, 
+AcademyAssist combines the power of a **Command Line Interface (CLI) with the user-friendly aspects of a Graphical
```
Comment (annabellekk) [inline docs/UserGuide.md:35]: Thanks for cleaning up the code!

File: src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        if (!arePrefixesPresent(argMultimap, PREFIX_FILTER_PARAM) || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, FilterCommand.MESSAGE_USAGE));
+        }
+
+        // Ensure that either the year group or subject prefix is present, but not both
+        boolean hasYearGroup = argMultimap.getValue(PREFIX_YEARGROUP).isPresent();
+        boolean hasSubject = argMultimap.getValue(PREFIX_SUBJECT).isPresent();
+
+        if (hasYearGroup && hasSubject) {
+            throw new ParseException("Please specify either a year group or a subject, not both.");
```
Comment (annabellekk) [inline src/main/java/seedu/academyassist/logic/parser/FilterCommandParser.java:42]: Good use of error message to help the user to trouble shoot

File: src/main/java/seedu/academyassist/logic/parser/ParserUtil.java [hunk 9]
```diff
@@ -97,7 +98,7 @@ public static Subject parseSubject(String subject) throws ParseException {
         if (!Subject.isValidSubject(trimmedSubject)) {
             throw new ParseException(Subject.MESSAGE_CONSTRAINTS);
         }
-        return new Subject(trimmedSubject);
+        return new Subject(trimmedSubject.toUpperCase());
```
Comment (annabellekk) [inline src/main/java/seedu/academyassist/logic/parser/ParserUtil.java:101]: Good that you consider different cases in which the input may be of different casing

File: src/main/java/seedu/academyassist/model/filter/FilterParam.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Returns True if FilterField of FilterParam is of type YEARGROUP
+     */
+    public boolean isYearGroup() {
+        return this.field.equals(FilterField.YEARGROUP);
+    }
+
+    /**
```
Comment (annabellekk) [inline src/main/java/seedu/academyassist/model/filter/FilterParam.java:50]: Good that you include javadoc comments for all the public methods to aid understanding and documentation

File: src/main/java/seedu/academyassist/ui/HelpWindow.java [hunk 11]
```diff
@@ -39,6 +39,8 @@ public class HelpWindow extends UiPart<Stage> {
             + "   Format: clear\n\n"
             + "9. exit: Exits the program\n"
             + "   Format: exit\n\n"
+            + "10. filter: Filters contacts based on specified field\n"
```
Comment (annabellekk) [inline src/main/java/seedu/academyassist/ui/HelpWindow.java:42]: Thanks for updating the help window also!

**Review by annabellekk — COMMENTED** [review]
Good work on the filter command!

**Pull request merged**